### PR TITLE
Use high performance session during gameplay on desktop platforms

### LINKED ISSUE #27306: Repeatable and consistent microstutters
### Type

Performance

### Bug description

This is a very strange issue. I've disabled everything I could and reinstalled a bunch of drivers, but I still get these weird and regular microstutters which bring an insanely annoying latency spike. These are the things that I managed to deduce:
1.) During gameplay, this stutter always happens in ~14 second intervals. Regardless of beatmap. I can also sometimes see the same graph spike when I restart a beatmap, although sometimes it doesnt appear. But it will always appear in the first ~9 seconds of a beatmap. After that its 14 seconds again.
2.) If im just idling in the song selection screen, I also get a consistent bump in the metrics every ~19 seconds. Its barely noticable in the graph, but it does have a bump in the Invalidations, Refreshes and DrawNodeAppl columns for Update and UniformUpl for Draw (there is absolutely no impact in Input and Audio).

Im seriously lost with this one. Tried using procmon to and task manager to see any CPU usage spikes, but there was nothing. Looking for any suggestions.

### Screenshots or videos

Spike

### Version

2024.221.0-lazer

### Logs

[1708561920.network.log](https://github.com/ppy/osu/files/14367193/1708561920.network.log)
[1708561920.runtime.log](https://github.com/ppy/osu/files/14367194/1708561920.runtime.log)
[1708561920.updater.log](https://github.com/ppy/osu/files/14367195/1708561920.updater.log)
[1708561920.performance.log](https://github.com/ppy/osu/files/14367196/1708561920.performance.log)

### PR DESCRIPTION
**Pull request by smoogipoo**
This was originally implemented in https://github.com/ppy/osu/pull/13525 and subsequently reverted in https://github.com/ppy/osu/pull/13591 because players experienced stutters.

`SustainedLowLatency` (the `GameHost` default) allows the GC to perform minimal gen0/gen1 collections, instead grouping them into larger batches. `LowLatency` instead significantly reduces gen0 size so that collections occur more frequently.  
A sub effect of this is that objects that survive gen0 are promoted to gen1, and collections there can take just as long as gen0. This is why it's not good to use LowLatency when a lot of alloc is being done, such as in SongSelect.

It's a balancing game, which is why I've put effort in this PR to make sure that the mode is only enabled during gameplay, and that the switch to it is invisible to the user (when the screen is mostly static). 

Players previously experienced stutters because the class relied on `LocalUserPlaying`, which only becomes `true` at the first hitobject's start time. A little bit too late.

Closes https://github.com/ppy/osu/issues/27306#issuecomment-1962722089

### CODE REVIEW

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -792,7 +792,7 @@ public override Task Import(ImportTask[] imports, ImportParameters parameters =
 
         protected virtual UpdateManager CreateUpdateManager() => new UpdateManager();
 
-        protected virtual HighPerformanceSession CreateHighPerformanceSession() => new HighPerformanceSession();
+        protected virtual HighPerformanceSessionManager CreateHighPerformanceSessionManager() => new HighPerformanceSessionManager();
```
Comment (peppy) [inline osu.Game/OsuGame.cs:795]: Was there a reason for this being `virtual`? Mobile or something? If not required, then:

```diff
diff --git a/osu.Game/OsuGame.cs b/osu.Game/OsuGame.cs
index f23a78d2dc..13e80e0707 100644
--- a/osu.Game/OsuGame.cs
+++ b/osu.Game/OsuGame.cs
@@ -792,8 +792,6 @@ public override Task Import(ImportTask[] imports, ImportParameters parameters =
 
         protected virtual UpdateManager CreateUpdateManager() => new UpdateManager();
 
-        protected virtual HighPerformanceSessionManager CreateHighPerformanceSessionManager() => new HighPerformanceSessionManager();
-
         protected override Container CreateScalingContainer() => new ScalingContainer(ScalingMode.Everything);
 
         #region Beatmap progression
@@ -1088,7 +1086,7 @@ protected override void LoadComplete()
             loadComponentSingleFile(new AccountCreationOverlay(), topMostOverlayContent.Add, true);
             loadComponentSingleFile<IDialogOverlay>(new DialogOverlay(), topMostOverlayContent.Add, true);
 
-            loadComponentSingleFile(CreateHighPerformanceSessionManager(), Add, true);
+            loadComponentSingleFile(new HighPerformanceSessionManager(), Add, true);
 
             loadComponentSingleFile(new BackgroundDataStoreProcessor(), Add);
 
diff --git a/osu.Game/Performance/HighPerformanceSessionManager.cs b/osu.Game/Performance/HighPerformanceSessionManager.cs
index c5146933b7..22c929083c 100644
--- a/osu.Game/Performance/HighPerformanceSessionManager.cs
+++ b/osu.Game/Performance/HighPerformanceSessionManager.cs
@@ -14,11 +14,11 @@ public partial class HighPerformanceSessionManager : Component
 
         public IDisposable BeginSession()
         {
-            EnableHighPerformanceSession();
-            return new InvokeOnDisposal<HighPerformanceSessionManager>(this, static m => m.DisableHighPerformanceSession());
+            enableHighPerformanceSession();
+            return new InvokeOnDisposal<HighPerformanceSessionManager>(this, static m => m.disableHighPerformanceSession());
         }
 
-        protected virtual void EnableHighPerformanceSession()
+        private void enableHighPerformanceSession()
         {
             originalGCMode = GCSettings.LatencyMode;
             GCSettings.LatencyMode = GCLatencyMode.LowLatency;
@@ -27,7 +27,7 @@ protected virtual void EnableHighPerformanceSession()
             GC.Collect(0);
         }
 
-        protected virtual void DisableHighPerformanceSession()
+        private void disableHighPerformanceSession()
         {
             if (GCSettings.LatencyMode == GCLatencyMode.LowLatency)
                 GCSettings.LatencyMode = originalGCMode;

```

Comment (peppy) [reply osu.Game/OsuGame.cs]: as long as it's not an `OperationNotSupported` 😅

**Review by bdach — CHANGES_REQUESTED** [review]
Dies on android:

```
2024-02-27 08:49:39 [error]: An unhandled error has occurred.
2024-02-27 08:49:39 [error]: System.PlatformNotSupportedException: Operation is not supported on this platform.
2024-02-27 08:49:39 [error]: at System.Runtime.GCSettings.SetGCLatencyMode(GCLatencyMode newLatencyMode)
2024-02-27 08:49:39 [error]: at System.Runtime.GCSettings.set_LatencyMode(GCLatencyMode value)
2024-02-27 08:49:39 [error]: at osu.Game.Performance.HighPerformanceSessionManager.enableHighPerformanceSession()
2024-02-27 08:49:39 [error]: at osu.Game.Performance.HighPerformanceSessionManager.BeginSession()
2024-02-27 08:49:39 [error]: at osu.Game.Screens.Play.PlayerLoader.pushWhenLoaded()
2024-02-27 08:49:39 [error]: at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
2024-02-27 08:49:39 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
2024-02-27 08:49:39 [error]: at osu.Framework.Threading.Scheduler.Update()
2024-02-27 08:49:39 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
```

File: osu.Game/Performance/HighPerformanceSessionManager.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            GC.Collect(0);
+        }
+
+        private void disableHighPerformanceSession()
+        {
+            if (GCSettings.LatencyMode == GCLatencyMode.LowLatency)
+                GCSettings.LatencyMode = originalGCMode;
+
+            // No GC.Collect() as we were already collecting at a higher frequency in the old mode.
+        }
```
Comment (bdach) [inline osu.Game/Performance/HighPerformanceSessionManager.cs:36]: I'd like to see logging in these methods too to be able to tell when this is being toggled.

**Review by peppy — APPROVED** [review]

**Review by bdach — APPROVED** [review]
let's see how this behaves in the wild this time round i guess

**Pull request merged**